mhcflurry_cloud/cross_validation.py

```diff
@@ -0,0 +1,159 @@
+from __future__ import absolute_import
+
+import collections
+import logging
+
+import pepdata
+
+from .common import AlleleSpecificTrainTestFold
+from .train_and_test import task_impute
+
+gbmr4_transformer = pepdata.reduced_alphabet.make_alphabet_transformer("gbmr4")
+def default_projector(peptide):
+    """
+    Given a peptide, return a list of projections for it. The projections are:
+        - the gbmr4 reduced representation
+        - for all positions in the peptide, the peptide with a "." replacing
+          the residue at that position
+
+    Peptides with overlapping projections are considered similar when doing
+    cross validation.
+
+    Parameters
+    ----------
+    peptide : string
+
+    Returns
+    ----------
+    string list
+    """
+    def projections(peptide, edit_distance=1):
+        if edit_distance == 0:
+            return set([peptide])
+        return set.union(*[
+                projections(p, edit_distance - 1)
+                for p in (
+                    peptide[0:i] + "." + peptide[(i + 1):]
+                    for i in range(len(peptide)))
+        ])
+    return sorted(projections(peptide)) + [gbmr4_transformer(peptide)]
+
+def similar_peptides(set1, set2, projector=default_projector):
+    """
+    Given two sets of peptides, return a list of the peptides whose reduced
+    representations are found in both sets.
+    
+    Parameters
+    ----------
+    projector : (string -> string) or (string -> string list)
+        Function giving projection(s) of a peptide
+        
+    Returns
+    ----------
+    string list
+        
+    """
+    result = collections.defaultdict(lambda: ([], []))
+    for (index, peptides) in enumerate([set1, set2]):
+        for peptide in peptides:
+            projections = projector(peptide)
+            if not isinstance(projections, list):
+                projections = [projections]
+            for projection in projections:
+                result[projection][index].append(peptide)
+    
+    common = set()
+    for (peptides1, peptides2) in result.values():
+        if peptides1 and peptides2:
+            common.update(peptides1 + peptides2)
+
+    return sorted(common)
+
+def cross_validation_folds(
+        train_data,
+        alleles=None,
+        n_folds=3,
+        drop_similar_peptides=False,
+        imputer=None,
+        impute_kwargs={
+            'min_observations_per_peptide': 2,
+            'min_observations_per_allele': 2,
+        }):
```

Comment:
Since there's nothing specific to parallelization here, maybe it should be eventually moved to mhcflurry?